common/test/client.go

```diff
@@ -82,7 +82,14 @@ func (r *Request) Do() (err error) {
 	if err != nil {
 		return err
 	}
-	defer (func() { err = res.Body.Close() })()
+	defer (func() {
```

Comment:
Is the function to close the http body called at all if in case of error it is returned before?
SEE: https://github.com/matrix-org/dendrite/pull/858/files#diff-8089a15307bfa30044c79197979a773bR82

Reply:
You're right. We should move the defer to right after the request is made. How about that?

Reply:
If an error is returned to do **not** need to close `res.Body`. The docs are very clear on this:

> On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned Response.Body is already closed. 

By putting it above the `err != nil` check, you risk cases where `err != nil && res == nil` resulting in a panic e.g on network errors.